client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -66,54 +87,64 @@
         Preconditions.checkNotNull(serializer);
         this.stream = stream;
         this.controller = controller;
+        this.segmentSealedCallBack = this::handleLogSealed;
         this.outputStreamFactory = outputStreamFactory;
         this.selector = new SegmentSelector(stream, controller, outputStreamFactory);
         this.serializer = serializer;
         this.config = config;
-        List<PendingEvent> failedEvents = selector.refreshSegmentEventWriters();
+        List<PendingEvent> failedEvents = selector.refreshSegmentEventWriters(segmentSealedCallBack);
         assert failedEvents.isEmpty() : "There should not be any events to have failed";
     }
 
     @Override
-    public AckFuture writeEvent(Type event) {
+    public CompletableFuture<Void> writeEvent(Type event) {
         return writeEventInternal(null, event);
     }
 
     @Override
-    public AckFuture writeEvent(String routingKey, Type event) {
+    public CompletableFuture<Void> writeEvent(String routingKey, Type event) {
         Preconditions.checkNotNull(routingKey);
         return writeEventInternal(routingKey, event);
     }
     
-    private AckFuture writeEventInternal(String routingKey, Type event) {
+    private CompletableFuture<Void> writeEventInternal(String routingKey, Type event) {
         Preconditions.checkNotNull(event);
         Exceptions.checkNotClosed(closed.get(), this);
         ByteBuffer data = serializer.serialize(event);
-        CompletableFuture<Boolean> result = new CompletableFuture<Boolean>();
-        synchronized (lock) {
+        CompletableFuture<Boolean> ackFuture = new CompletableFuture<Boolean>();
+        lock.writeLock().lock();
```

Comment:
Is this serializing all writes independent of key, would it make sense to make the locking more fine grained so that we don't block all writes when we don't have a writer for a specific segment?